File: Solid/nn/plotPCA.jl
using NPZ
using LinearAlgebra
using PyPlot
include("../../plotdefaults.jl")

using Random, Distributions
include("../src/PlatePull.jl")


DeepONetPCA_data = npzread("DeepONet/DeepONetPCA_data.npy")
Ug = npzread("PCA/PCA_data.npy")
    

porder = 2
θ = rand(Normal(0, 1.0), 100);
filename = "../src/square-circle-coarse-o2"
domain, Fn = ConstructDomain(porder, θ, filename)
ngp = Int64(sqrt(length(domain.elements[1].weights)))

# fig2,ax = PyPlot.subplots(3,4,sharex=true,figsize = (6.5,5))
fig2 = PyPlot.figure(figsize=(6,5))
subfigs = fig2.subfigures(nrows=3,ncols=1)

ax1 = subfigs[1].subplots(1,4)
ax2 = subfigs[2].subplots(1,4)
ax3 = subfigs[3].subplots(1,4)
ax = [ax1, ax2, ax3]
ims = Array{Any}(undef,3,4)
for i = 1:4
    ims[1,i] = visσ(domain, ngp, -0.1, 0.1; σ=Ug[:, i], ax = ax1[i], mycolorbar="gray")
    ims[2,i] = visσ(domain, ngp,0,0.15; σ=DeepONetPCA_data[:, 2*i-1], ax = ax2[i], mycolorbar="gray")
    ims[3,i] = visσ(domain, ngp,-0.05,0.1; σ=DeepONetPCA_data[:, 2*i], ax = ax3[i], mycolorbar="gray")
    # @show minimum(DeepONetPCA_data[:, 2*i]),maximum(DeepONetPCA_data[:, 2*i])

    for j = 1:3
        ax[j][i].spines["top"].set_visible(false)
        ax[j][i].spines["right"].set_visible(false)
        ax[j][i].spines["left"].set_visible(false)
        ax[j][i].spines["bottom"].set_visible(false)
        ax[j][i].set_aspect("equal", "box")
        ax[j][i].set_yticks([])
        ax[j][i].set_xticks([])
    end
end

subfigs[1].suptitle("Leading PCA basis functions",fontsize=14,y=1)
subfigs[2].suptitle("Trained DeepONet trunk functions",fontsize=14,y=1)
subfigs[3].suptitle("Leading PCA modes of trained DeepONet trunk functions",fontsize=14,y=1)
plt.subplots_adjust(left = 0.01, right = 0.9, bottom = 0.025,top=.95,hspace=0.2,wspace=0.1)

i = 4
temp = ax[1][i].get_position()
xw = temp.x1-temp.x0
cax = subfigs[1].add_axes([temp.x1+0.03*xw, temp.y0, 0.07*xw, temp.y1-temp.y0],frameon=false)
cb = plt.colorbar(ims[1,i],cax=cax)
cb.outline.set_visible(false)
cb.ax.yaxis.set_tick_params(colors="#808080",width=0.3)

temp = ax[2][i].get_position()
xw = temp.x1-temp.x0
cax = subfigs[2].add_axes([temp.x1+0.03*xw, temp.y0, 0.07*xw, temp.y1-temp.y0],frameon=false)
cb = plt.colorbar(ims[2,i],cax=cax)
cb.outline.set_visible(false)
cb.ax.yaxis.set_tick_params(colors="#808080",width=0.3)

temp = ax[3][i].get_position()
xw = temp.x1-temp.x0
cax = subfigs[3].add_axes([temp.x1+0.03*xw, temp.y0, 0.07*xw, temp.y1-temp.y0],frameon=false)
cb = plt.colorbar(ims[3,i],cax=cax)
cb.outline.set_visible(false)
cb.ax.yaxis.set_tick_params(colors="#808080",width=0.3)


fig2.savefig("Solid-pca-vis.pdf")
plt.close()

File: Helmholtz/nn/plot_pca.py
import sys
import numpy as np
sys.path.append('../../../nn')
from datetime import datetime

import matplotlib as mpl 
from matplotlib.lines import Line2D 
# mpl.use('TkAgg')
import matplotlib.pyplot as plt

plt.rc("figure", dpi=300)           # High-quality figure ("dots-per-inch")
plt.rc("text", usetex=True)         # Crisp axis ticks
plt.rc("font", family="serif")      # Crisp axis labels
plt.rc("legend", edgecolor='none')  # No boxes around legends

plt.rc("figure",facecolor="#ffffff")
plt.rc("axes",facecolor="#ffffff",edgecolor="#808080",labelcolor="#000000")
plt.rc("savefig",facecolor="#ffffff")
plt.rc("text",color="#000000")
plt.rc("xtick",color="#808080")
plt.rc("ytick",color="#808080")

#######################################
#
########################################
M = 20000
N = 101

recomputePCA = False
if recomputePCA:
    ntrain = M//2
    N_theta = 100
    prefix = "../../data/"
    # theta = np.load(prefix+"Random_NS_theta_" + str(N_theta) + ".npy")   
    outputs = np.load(prefix+"Random_Helmholtz_high_K_" + str(N_theta) + ".npy")
    # inputs = np.load(prefix+"Random_Helmholtz_high_cs_" + str(N_theta) + ".npy")

    acc=0.999


    train_outputs = np.reshape(outputs[:,:,:M//2], (-1, M//2))
    test_outputs  = np.reshape(outputs[:,:,M//2:M], (-1, M-M//2))
    Uo,So,Vo = np.linalg.svd(train_outputs)
    en_g = 1 - np.cumsum(So)/np.sum(So)
    r_g = np.argwhere(en_g<(1-acc))[0,0]
    Ug = Uo[:,:r_g]
    np.save("HHpca.npy",Ug[:,:4])
else:
    Ug = np.load("HHpca.npy")

DeepONetPCA_data = np.load("DeepONet/DeepONetPCA_data.npy")
xgrid = np.linspace(0,1,N)
Y, X = np.meshgrid(xgrid, xgrid)

# fig,ax = plt.subplots(3,4,sharex=True,sharey=True,figsize = (6.5,5.5))
fig = plt.figure(figsize=(6.5,5.5))
subfigs = fig.subfigures(nrows=3,ncols=1)
axs = []
for row,subfig in enumerate(subfigs):
    axs.append(subfig.subplots(1,4))
ims = []
for i in range(4):
    ims.append(axs[0][i].pcolormesh(X,Y,np.reshape(Ug[:, i], (N,N)),shading="gouraud",cmap="gray",vmin=-0.042,vmax=0.042))
    ims.append(axs[1][i].pcolormesh(X, Y, DeepONetPCA_data[:,:,2*i], shading="gouraud",cmap="gray",vmin=0,vmax=0.1))
    ims.append(axs[2][i].pcolormesh(X,Y,DeepONetPCA_data[:,:,2*i+1],shading="gouraud",cmap="gray",vmin=-0.04,vmax=0.04))
    print(np.min(DeepONetPCA_data[:,:,2*i+1]),np.max(DeepONetPCA_data[:,:,2*i+1]))

    for j in range(3):
        axs[j][i].set_aspect("equal","box")
        axs[j][i].set_axis_off()
    
subfigs[0].suptitle("Leading PCA basis functions",fontsize=16,y=0.95)
subfigs[1].suptitle("Trained DeepONet trunk functions",fontsize=16,y=0.95)
subfigs[2].suptitle("Leading PCA modes of trained DeepONet trunk functions",fontsize=16,y=0.95)

plt.subplots_adjust(left=0.02,right=0.87,bottom=0.02,top=0.90)
cax = []
for i in range(3):
    temp = axs[i][3].get_position()
    cax.append(subfigs[i].add_axes([0.9,temp.y0,0.02,temp.y1-temp.y0]))
    if i==0:
        cb = plt.colorbar(ims[i],cax=cax[i],ticks=[-0.04,0,0.04])
    elif i == 1:
        cb = plt.colorbar(ims[i],cax=cax[i],ticks=[0,0.05,0.1])
    else:
        cb = plt.colorbar(ims[i],cax=cax[i],ticks=[-0.04,0,0.04])
    cb.outline.set_visible(False)
    cb.ax.yaxis.set_tick_params(width=0.3)
    for t in cb.ax.get_yticklabels():
        t.set_fontsize(12)

fig.savefig("Helmholtz-pca-vis.pdf")
plt.close("all")
# plt.colorbar()
